### ISSUE TITLE
FreeBSD compilation error

### ISSUE THREAD

**Issue by tymat** [briansmith/ring#311]
Version Info:
```
rustc 1.12.0
cargo 0.13.0 (built 2016-10-06)
FreeBSD 11.0-RELEASE-p1 FreeBSD 11.0-RELEASE-p1 #0 r306420: Thu Sep 29 01:43:23 UTC 2016     root@releng2.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
```

```
RUST_BACKTRACE=1 cargo build
   Compiling ring v0.4.3 (file:///usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3)
error: failed to run custom build command for `ring v0.4.3 (file:///usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3)`
process didn't exit successfully: `/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/build-script-build` (exit code: 101)
--- stdout
SSL_CERT_DIR: /usr/local/share/certs
SSL_CERT_FILE: /usr/local/openssl/cert.pem
RUST_BACKTRACE: 1
_system_type: BSD
SSH_CONNECTION: 50.35.93.68 63545 46.101.95.207 22
LANG: en_US.UTF-8
rvm_bin_path: /home/tymat/.rvm/bin
OLDPWD: /home/tymat/parity
EDITOR: vi
rvm_version: 1.27.0 (master)
RUBY_VERSION: ruby-2.3.1
MM_CHARSET: UTF-8
GEM_HOME: /home/tymat/.rvm/gems/ruby-2.3.1
USER: tymat
ENV: /home/tymat/.shrc
PAGER: more
PWD: /usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3
_system_version: 11.0
HOME: /home/tymat
_system_name: FreeBSD
SSH_CLIENT: 50.35.93.68 63545 22
_system_arch: x86_64
GEM_PATH: /home/tymat/.rvm/gems/ruby-2.3.1:/home/tymat/.rvm/gems/ruby-2.3.1@global
SSH_TTY: /dev/pts/0
rvm_path: /home/tymat/.rvm
MAIL: /var/mail/tymat
SHELL: /usr/local/bin/bash
TERM: xterm-256color
rvm_prefix: /home/tymat
SHLVL: 1
BLOCKSIZE: K
LOGNAME: tymat
MY_RUBY_HOME: /home/tymat/.rvm/rubies/ruby-2.3.1
PATH: /home/tymat/.rvm/gems/ruby-2.3.1/bin:/home/tymat/.rvm/gems/ruby-2.3.1@global/bin:/home/tymat/.rvm/rubies/ruby-2.3.1/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/home/tymat/bin:/home/tymat/.rvm/bin:/home/tymat/.rvm/bin:/home/tymat/.rvm/bin
IRBRC: /home/tymat/.rvm/rubies/ruby-2.3.1/.irbrc
_: /usr/local/bin/cargo
CARGO_PKG_AUTHORS: Brian Smith <brian@briansmith.org>
CARGO_PKG_HOMEPAGE:
PROFILE: debug
CARGO_PKG_NAME: ring
DEBUG: true
CARGO_FEATURE_DEV_URANDOM_FALLBACK: 1
CARGO_PKG_VERSION: 0.4.3
CARGO_FEATURE_DEFAULT: 1
CARGO_FEATURE_USE_HEAP: 1
RUSTC: rustc
CARGO_PKG_VERSION_MAJOR: 0
RUSTDOC: rustdoc
CARGO_PKG_VERSION_PATCH: 3
CARGO_PKG_VERSION_PRE:
TARGET: x86_64-unknown-freebsd
LD_LIBRARY_PATH: /usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/deps
OPT_LEVEL: 0
CARGO_PKG_VERSION_MINOR: 4
NUM_JOBS: 2
CARGO_MANIFEST_DIR: /usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3
HOST: x86_64-unknown-freebsd
OUT_DIR: /usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out
CARGO_PKG_DESCRIPTION: Safe, fast, small crypto using Rust.

make: stopped in /usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3

--- stderr
make: "mk/mk/top_of_makefile.mk" line 35: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 37: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 40: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 44: Need an operator
make: "mk/mk/top_of_makefile.mk" line 45: Need an operator
make: "mk/mk/top_of_makefile.mk" line 48: Need an operator
make: "mk/mk/top_of_makefile.mk" line 58: Need an operator
make: "mk/mk/top_of_makefile.mk" line 59: Need an operator
make: "mk/mk/top_of_makefile.mk" line 60: Need an operator
make: "mk/mk/top_of_makefile.mk" line 64: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 65: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 67: Need an operator
make: "mk/mk/top_of_makefile.mk" line 68: Need an operator
make: "mk/mk/top_of_makefile.mk" line 70: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 73: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 76: Need an operator
make: "mk/mk/top_of_makefile.mk" line 80: Need an operator
make: "mk/mk/top_of_makefile.mk" line 82: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 84: Need an operator
make: "mk/mk/top_of_makefile.mk" line 89: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 91: Need an operator
make: "mk/mk/top_of_makefile.mk" line 101: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 102: Need an operator
make: "mk/mk/top_of_makefile.mk" line 103: Need an operator
make: "mk/mk/top_of_makefile.mk" line 104: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 105: Need an operator
make: "mk/mk/top_of_makefile.mk" line 106: Need an operator
make: "mk/mk/top_of_makefile.mk" line 128: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 130: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 133: Need an operator
make: "mk/mk/top_of_makefile.mk" line 136: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 139: Need an operator
make: "mk/mk/top_of_makefile.mk" line 142: Need an operator
make: "mk/mk/top_of_makefile.mk" line 143: Need an operator
make: "mk/mk/top_of_makefile.mk" line 147: Need an operator
make: "mk/mk/top_of_makefile.mk" line 151: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 155: Need an operator
make: "mk/mk/top_of_makefile.mk" line 158: Need an operator
make: "mk/mk/top_of_makefile.mk" line 222: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 224: Need an operator
make: "mk/mk/top_of_makefile.mk" line 243: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 245: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 247: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 249: Missing dependency operator
make: "mk/mk/top_of_makefile.mk" line 252: Need an operator
make: Unknown modifier ' '
make: "mk/mk/top_of_makefile.mk" line 253: Need an operator
make: "mk/mk/top_of_makefile.mk" line 254: Need an operator
make: "mk/mk/ring.mk" line 202: Missing dependency operator
make: "mk/mk/ring.mk" line 204: Missing dependency operator
make: "mk/mk/ring.mk" line 205: Missing dependency operator
make: "mk/mk/ring.mk" line 207: Missing dependency operator
make: "mk/mk/ring.mk" line 209: Need an operator
make: "mk/mk/ring.mk" line 211: Need an operator
make: "mk/mk/ring.mk" line 212: Missing dependency operator
make: "mk/mk/ring.mk" line 214: Missing dependency operator
make: "mk/mk/ring.mk" line 216: Need an operator
make: "mk/mk/ring.mk" line 218: Need an operator
make: "mk/mk/bottom_of_makefile.mk" line 37: warning: duplicate script for target "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" ignored
make: "mk/mk/bottom_of_makefile.mk" line 35: warning: using previous script for "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" defined here
make: "mk/mk/bottom_of_makefile.mk" line 39: warning: duplicate script for target "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" ignored
make: "mk/mk/bottom_of_makefile.mk" line 35: warning: using previous script for "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" defined here
make: "mk/mk/bottom_of_makefile.mk" line 41: warning: duplicate script for target "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" ignored
make: "mk/mk/bottom_of_makefile.mk" line 35: warning: using previous script for "/usr/home/tymat/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-0.4.3/target/debug/build/ring-c1fa8cfa8b1e39a6/out/obj/%.o" defined here
make: "mk/mk/bottom_of_makefile.mk" line 54: Missing dependency operator
make: Fatal errors encountered -- cannot continuethread 'main' panicked at 'make execution failed', build.rs:211
stack backtrace:
   1:          0x1044c79 - <unknown>
   2:          0x1049d8c - <unknown>
   3:          0x1049232 - <unknown>
   4:          0x10498c4 - <unknown>
   5:          0x1049722 - <unknown>
   6:          0x1049690 - <unknown>
   7:          0x103abcd - <unknown>
   8:          0x1038a23 - <unknown>
   9:          0x10379c9 - <unknown>
  10:          0x1054246 - <unknown>
  11:          0x1048a56 - <unknown>
  12:          0x103b153 - <unknown>
  13:          0x102a3f0 - <unknown>
  14:        0x8012b3fff - <unknown>
```

**Comment by tatsuya6502** [comment]
I just ran into the same problem (rustc 1.12.0, FreeBSD 10.3-RELEASE-p7 amd64). I believe ring does not support building on FreeBSD yet.

So far, I found two places to fix:
- [build.rs#L116](https://github.com/briansmith/ring/blob/5a2c0a2006165db9da3721c4b49300ded4dacd5b/build.rs#L116): On FreeBSD, `gmake` should be used instead of `make` .
- [mk/top_of_makefile.mk](https://github.com/briansmith/ring/blob/1f020f30a3eb980487dca7ebcf1a369421134a7a/mk/top_of_makefile.mk): `TARGET: x86_64-unknown-freebsd` should be supported. 

I will try to find some spare time to fix them.

**Comment by tatsuya6502** [comment]
Relevant issue:
- [Cross compilation on *BSD #287](https://github.com/briansmith/ring/issues/287)

**Comment by tatsuya6502** [comment]
@tymat I opened a PR #313 to address this issue. You can give it a try by having the followings in your `Cargo.toml`.

```toml
[dependencies]
ring = { git = "https://github.com/tatsuya6502/ring/", rev = "bsd-support" }
```

**Comment by tymat** [comment]
@tatsuya6502   thanks.  This fix works.

**Comment by briansmith** [comment]
Fixed via #313.
